### PR TITLE
explain: don't swallow regular tracing spans and events

### PR DESCRIPTION
**Pull request by aalexandrov**
Fixes an issue that was blocking @mgree last Friday.

### Motivation

   * This PR refactors existing code.

The current `OptimizerTrace` implementation substitutes the default tracing `Subscriber` with a brand new one containing only `PlanTrace` layers. This means that regular `tracing::event!` calls are not propagated to the regularly configured subscribers, which can be a problem if one tries to rely on them for debugging code that lives inside `explain_query_optimizer_pipeline`.

This PR fixes that limitation by introducing a `DelegateSubscriber` that can be used to delegate subscriber calls to an inner subscriber. The new subscriber type is used as the base of the `OptimizerTrace` subscriber chain instead of `tracing_subscriber::registry()` when you type `EXPLAIN ... BROKEN <query>`.

**Note**: Previous iterations of this PR have revealed that installing a `DelegateSubscriber` leaks memory. The use of `EXPLAIN ... BROKEN <query>` therefore is put behind a feature flag (`enable_explain_broken`) so we don't hit this leak in production (also, the syntax is only meant to be used internally when debugging optimizer pipeline panics).

Developers that want to use this syntax will first need to run this command on 6877 as `mz_system`:

```sql
alter system set enable_explain_broken = true;
```

### Tips for reviewer

It's hard to add a test for this, but I validated manually that `tracing::info!` calls in `explain_query_optimizer_pipeline` can be now seen in the regular logs.

@guswynn / @pH14 Can you please check the changes of the second commit at a high level and voice any concerns of the adopted approach. It seems that when I install a `DelegateSubscriber` I start leaking memory somehow. Regartless, I think it should be safe to merge this because the syntax that triggers that behavior is now disabled by default.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by def- — CHANGES_REQUESTED** [review]
As discussed on Slack (https://materializeinc.slack.com/archives/C01LKF361MZ/p1695133172107989), there seems to be a memory leak when running explains against this PR using SQLsmith. I'm so far unable to reproduce it with explain queries on main and non-explain queries on main either, so not sure if it's related to https://github.com/MaterializeInc/materialize/issues/21683.

**Review by def- — APPROVED** [review]
Not seeing increased environmentd memory usage with SQLsmith locally with the new change adding "broken".

**Comment by aalexandrov** [comment]
@mgree if you are still blocked on debugging async code in the `explain_query_optimizer_pipeline` you can rebase on this commit and do:

```sql
ALTER SYSTEM SET enable_explain_broken = true;
```

as `mz_system` on 6877 and then run

```sql
EXPLAIN ... PLAN FOR BROKEN <query>
```

with your instrumented code.

**Review by pH14 — COMMENTED** [review]
Didn't do a line-level review of the subscriber changes, but at a high level it seems quite reasonable (and very clever!)

File: src/repr/src/explain/tracing.rs [hunk 1]
```diff
@@ -315,6 +319,62 @@ impl field::Visit for ExtractStr {
...
+pub struct DelegateSubscriber {
+    inner: Dispatch,
+}
+
+impl Default for DelegateSubscriber {
+    fn default() -> Self {
+        let default = tracing::dispatcher::get_default(Clone::clone);
+
+        let inner = if default.downcast_ref::<Registry>().is_none() {
+            Registry::default().into() // we need at least a simple registry
```
Comment (guswynn) [inline src/repr/src/explain/tracing.rs:335]: when is this case actually hit?

**Review by guswynn — COMMENTED** [review]
I believe the leak is the reference cycles between the stored `Dispatch` in `DelegateSubscriber` and the one registered with the future, but i'm not 100% sure about how that works

Because its behind a flag, this is fine to merge, but I think a better technique would be to:

- pre-coerce our default global subscriber into an `Arc<dyn Subscriber>` (this already happens when its set as the global default), and store it in a global we can clone
- stack the `PlanTrace` layers on top of (using `.with_filter` to ensure they only see the relevant traces).
- extend our fmt filters to exclude the `optimizer` target, so we can leave them at `info!` level. This ensure the tracing fast path still works (we can ensure this is tested in `test/tracing`)

I think this avoids the leak? It also avoids us implementing `Subscriber`, which is not really possible to do 100% correctly (there are hidden methods)

**Comment by aalexandrov** [comment]
> I believe the leak is the reference cycles between the stored `Dispatch` in `DelegateSubscriber` and the one registered with the future, but i'm not 100% sure about how that works
> 
> Because its behind a flag, this is fine to merge, but I think a better technique would be to:
> 

Thanks, @guswynn, I'll try this out as a follow-up.

> * pre-coerce our default global subscriber into an `Arc<dyn Subscriber>` (this already happens when its set as the global default), and store it in a global we can clone

I guess this needs to happen in `mz_ore`, right?

> * stack the `PlanTrace` layers on top of (using `.with_filter` to ensure they only see the relevant traces).

From the `Filter` docs it seems that the `Filter` is only applied to the current layer, not to preceding wrapped layers in the current layer chain.

> * extend our fmt filters to exclude the `optimizer` target, so we can leave them at `info!` level. This ensure the tracing fast path still works (we can ensure this is tested in `test/tracing`)

I don't know if this helps, but the `target = optimizer` items that are relevant to optimizer tracing are only spans, not events. The `trace_plan<T>` method just tries to coerce the current `Subscriber` to a `PlanTrace<T>` and appends a `plan: T` to a vector owned by the `PlanTrace<T>` instance, so `PlanTrace<T>` doesn't actually require a `Registry` at the bottom of the layer chain.

> I think this avoids the leak? It also avoids us implementing `Subscriber`, which is not really possible to do 100% correctly (there are hidden methods)

I'll give it a shot and let you know.

File: src/repr/src/explain/tracing.rs (hunk 1 above)
Comment (aalexandrov) [reply src/repr/src/explain/tracing.rs:335]: I was just future-proofing against scenarios where this runs without an `mz_ore::tracing::configure()` call (this happens in `*.td` tests for example).

**Comment by aalexandrov** [comment]
> I believe the leak is the reference cycles between the stored Dispatch in DelegateSubscriber and the one registered with the future, but i'm not 100% sure about how that works

After doing some more reading that the memory leak is due to growing `CALLSITES` list after each `EXPLAIN` call.

> pre-coerce our default global subscriber into an Arc<dyn Subscriber> (this already happens when its set as the global default), and store it in a global we can clone

I tried this, but I'm not sure `tracing::dispatch` provides the right API to do that (at least I wasn't able to find a way to do that before the `stack.init()` call in `mz_ore`.

**Comment by aalexandrov** [comment]
I [asked in Discord](https://discord.com/channels/500028886025895936/627649734592561152/1155831166499303425) about possible ways to do this correctly.

**Comment by guswynn** [comment]
> From the Filter docs it seems that the Filter is only applied to the current layer, not to preceding wrapped layers in the current layer chain.

you can clone the filter and use it for all the layers! Or add a global `EnvFilter`

@aalexandrov curious to know what discord says, let me know if you want to pair program on this at some point!

**Comment by guswynn** [comment]
> I tried this, but I'm not sure tracing::dispatch provides the right API to do that (at least I wasn't able to find a way to do that before the stack.init() call in mz_ore.

hmm, i think it should https://docs.rs/tracing-core/0.1.30/tracing_core/subscriber/trait.Subscriber.html#impl-Subscriber-for-Arc%3CS%3E should work here?

**Comment by aalexandrov** [comment]
> you can clone the filter and use it for all the layers!

At the end I also implemented `Filter` for `PlanTrace` to hard-code the filter to be applied for the corresponding `Layer` implementation.

**Comment by aalexandrov** [comment]
> > I tried this, but I'm not sure tracing::dispatch provides the right API to do that (at least I wasn't able to find a way to do that before the stack.init() call in mz_ore.
> 
> hmm, i think it should [docs.rs/tracing-core/0.1.30/tracing_core/subscriber/trait.Subscriber.html#impl-Subscriber-for-Arc%3CS%3E](https://docs.rs/tracing-core/0.1.30/tracing_core/subscriber/trait.Subscriber.html#impl-Subscriber-for-Arc%3CS%3E) should work here?

This kind of works -- I can now initialize a static variable `STACK` of type `Arc<dyn Subscriber + Send + Sync>`. 

```rust
// in mz_ore::tracing
static STACK: OnceLock<Arc<dyn Subscriber + Sync + Send>> = OnceLock::new();

// in mz_ore::tracing::configure
STACK.get_or_init(|| Arc::new(stack));
```

However, when I call

```rust
STACK.get().expect("unwrap").clone().init();
```

I will hit [the `Dispatch::new` constructor](https://docs.rs/tracing/latest/tracing/dispatcher/struct.Dispatch.html#method.new) via [the blanket `From` implementation](https://docs.rs/tracing/latest/tracing/dispatcher/struct.Dispatch.html#impl-From%3CS%3E-for-Dispatch), and inside I will wrap the `Arc` into another `Arc`:

```rust
/// Returns a new `Dispatch` that discards events and spans.
#[inline]
pub fn none() -> Self {
    Dispatch {
        subscriber: Arc::new(NoSubscriber::default()),
    }
}
```

**Comment by guswynn** [comment]
@aalexandrov oh thats a tragedy :(

We could manually set the default and then copy this: https://docs.rs/tracing-subscriber/latest/src/tracing_subscriber/util.rs.html#66-71 block to do the same thing as `init`, but that requires `Dispatch` letting you init it with a `Arc<dyn Subscriber>`

that should be easy enough to add, I can do that

**Comment by guswynn** [comment]
@aalexandrov I suspect it will not be controversial, ill message eliza and the others in discord later this week

**Pull request merged**